### PR TITLE
Fix a bug in the Span where the returned options was not unmodifiable.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Remove forgotten method from a previous PR.
Add check for consistency between SpanContext sampled and RECORD_EVENTS.

### CODE REVIEW

File: core/src/main/java/com/google/instrumentation/trace/Span.java [hunk 1]
```diff
@@ -46,17 +48,25 @@
...
    *     options will be set.
    * @throws NullPointerException if context is {@code null}.
+   * @throws IllegalArgumentException is the {@code SpanContext} is sampled but no RECORD_EVENTS
+   *     options.
    */
   protected Span(SpanContext context, @Nullable EnumSet<Options> options) {
     this.context = checkNotNull(context, "context");
-    this.options = options == null ? EnumSet.noneOf(Options.class) : EnumSet.copyOf(options);
+    this.options =
+        Collections.unmodifiableSet(options == null ? DEFAULT_OPTIONS : EnumSet.copyOf(options));
```
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/trace/Span.java]: Suggestion: I think it's simpler to use Collections.unmodifiableSet to create DEFAULT_OPTIONS and then have:
```java
 this.options =
         options == null ? DEFAULT_OPTIONS : Collections.unmodifiableSet(EnumSet.copyOf(options));
```

File: core/src/main/java/com/google/instrumentation/trace/Span.java [hunk 2]
```diff
@@ -46,17 +48,25 @@
...
+  private static final EnumSet<Options> DEFAULT_OPTIONS = EnumSet.noneOf(Options.class);
+
   /**
    * Creates a new {@code Span}.
    *
    * @param context The context associated with this {@code Span}.
    * @param options The options associated with this {@code Span}. If {@code null} then default
    *     options will be set.
    * @throws NullPointerException if context is {@code null}.
+   * @throws IllegalArgumentException is the {@code SpanContext} is sampled but no RECORD_EVENTS
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/Span.java]: s/is/if

File: core/src/main/java/com/google/instrumentation/trace/Span.java [hunk 3]
```diff
@@ -46,17 +48,25 @@
...
+   *     options.
    */
   protected Span(SpanContext context, @Nullable EnumSet<Options> options) {
     this.context = checkNotNull(context, "context");
-    this.options = options == null ? EnumSet.noneOf(Options.class) : EnumSet.copyOf(options);
+    this.options =
+        Collections.unmodifiableSet(options == null ? DEFAULT_OPTIONS : EnumSet.copyOf(options));
+    checkArgument(
+        !context.getTraceOptions().isSampled() || (this.options.contains(Options.RECORD_EVENTS)),
+        "Span without record_events but sampled.");
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/trace/Span.java]: Think message would be slightly clearer as "Span is sampled, but does not have RECORD_EVENTS set."

File: core/src/main/java/com/google/instrumentation/trace/Span.java (hunk 1 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/Span.java]: Thanks.

File: core/src/main/java/com/google/instrumentation/trace/Span.java (hunk 2 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/Span.java]: Done.

File: core/src/main/java/com/google/instrumentation/trace/Span.java (hunk 3 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/trace/Span.java]: Done.

**Comment by bogdandrutu** [comment]
Added tests as well.

**Pull request merged**